api/spec/v1.yaml

```diff
@@ -1500,6 +1500,15 @@ components:
         message:
           type: string
           description: The message of a failed transaction.
+        params:
+          type: array
+          items:
+            $ref: '#/components/schemas/EvmAbiParam'
+          description: |
+            The error parameters, as decoded using the contract abi.
+            Only present for errors from verified contracts. Note that
+            users should be cautious when evaluating error data since the
+            data origin is not tracked and the error can be faked.
```

Comment:
IIRC technically the error is never (except at old heights) just a plain string; `message` is our stringification of a name-and-anonymous-param structure that's something like `Error("this is my error message")`. So in my mind, it doesn't make sense to expose `message` _and_ `params`; `message` is just a human-friendly rendition of `params` (if there are any; or just the plain text instead; and also possibly a non-evm prefix stirng like "reverted:").
Or maybe we want to argue that rendering the message is tricky (especially with the "reverted: " prefix and whatnot) even when structured params are given, and that pre-rendering into `message` is helpful, but it's also helpful for people to be able to access the structured part? I'm doubtful about the latter, but not strongly opposed to it.

The prior behavior on parsing errors, see RawMessage vs Message (= parsed RawMessage) here:
https://github.com/oasisprotocol/nexus/blob/6d23919d10ce55b8446ed3edddbb7ce4c7219545/analyzer/runtime/extract.go#L73-L77
You can see how the latter is derived from the former, and there are also unit tests.

Would it make sense for the backfilling parser to reuse more of the non-backfilling parser code? E.g. this function:
https://github.com/oasisprotocol/nexus/blob/6d23919d10ce55b8446ed3edddbb7ce4c7219545/analyzer/runtime/extract.go#L676-L676
It currently only attempts to parse with a single ABI, `Error(string)`, see line 698. The func would have to made a little more generic so the backfilling parser can pass in a custom ABI. Then we could use it in the backfilling analyzer.

At this point, I think I might be missing something. https://github.com/oasisprotocol/nexus/blob/6d23919d10ce55b8446ed3edddbb7ce4c7219545/analyzer/evmabibackfill/evm_abi_backfill.go#L162-L165 even mentions `tryParseErrorMessage()` (which I'm trying to get us to reuse here) but then does not call it. Maybe we can clarify in a chat or call?

Reply:
> expose message and params; message

Hm the `txErr.Message` field stores the error message for all types of tx errors. Non evm revert errors, such as `execution failed: out of gas`, are still returned as plaintext even in the most recent upgrade, so we can't remove the `message` field and replace it with `params`.

Agreed that the repetition between `message` and `params` for newer evm reverted txs is somewhat messy. We _could_ avoid this by not populating `params` when we specifically see that the abi error has the format `Error(string)`, but imo it's not worth special-casing this.

Edit**: Actually it looks like the `Error(string)` type is not included by default in the abi [due to quirks with evm](https://docs.soliditylang.org/en/latest/control-structures.html#revert). `reverted("my_error")` is automatically encoded as the `Error(string)` type but isn't included in the abi. The backfilling analyzer shouldn't be able to parse such errors since they're not explicitly defined in the abi. Therefore we should continue optimistically trying to parse tx evm reverted errors as `Error(string)`. One caveat: some contracts may explicitly define the type, which would lead to the error message being replicated in `message` and `params`, but we have no control over this and it should be rare.

> Would it make sense for the backfilling parser to reuse more of the non-backfilling parser code?

Yeah the explicit abi logic is a remnant from the days before Warren added the `abiparse` package. We should define a bare-bones `abi` struct with an `Error(string)` error and update `tryParseErrorMessage` to use the new `abiparse.ParseError(err, abi)` function. I'd prefer to do this in a follow-up PR; filed a ticket to track: https://app.clickup.com/t/8693epjgf

Reply:
Updated, thanks!

> `message` will contain the revert reason, either in its raw ABI-encoded form

We currently do not return the raw abi-encoded message; we only store it in `runtime_transactions.error_message_raw`. We could easily change the api to return the raw abi-encoded message if no parsed message is found, though I'm not sure how useful that would be for most users. Perhaps the main use case would be for contract devs who have the abi and are able to decode it themselves?